torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -26,11 +26,17 @@ PyObject* THPCppFunction_call(PyObject* self, PyObject* args, PyObject *kwargs)
     return PyErr_Format(PyExc_TypeError, "keyword arguments are not supported");
   }
 
+  auto& fn = *((THPCppFunction*)self);
   int num_inputs = PyTuple_GET_SIZE(args);
+  if (num_inputs != fn.num_args) {
+    return PyErr_Format(PyExc_TypeError, "expect %d arguments (got %d)",
+        fn.num_args, num_inputs);
+  }
+
   variable_list vars(num_inputs);
   for (int i = 0; i != num_inputs; ++i) {
     PyObject* arg = PyTuple_GET_ITEM(args, i);
-    if (arg == Py_None) {
+    if (i >= fn.num_required_args && arg == Py_None) {
```

Comment:
I think this might be incorrect in case there are two optional arguments, that don't need to be not None together. `(arg1, None, arg2)` will then look the same as `(arg1, arg2, None)`. Maybe let's only check the number and availability of args at the function level?